### PR TITLE
cop: add compare into new expression calculate framework.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

File: src/coprocessor/dag/expr/compare.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+                let rhs_unsigned = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+                let ordering = cmp_i64_with_unsigned_flag(lhs, lhs_unsigned, rhs, rhs_unsigned);
+                Ok(Some(match sig {
+                    ScalarFuncSig::LTInt => ordering == Ordering::Less,
+                    ScalarFuncSig::LEInt => ordering != Ordering::Greater,
+                    ScalarFuncSig::GTInt => ordering == Ordering::Greater,
+                    ScalarFuncSig::GEInt => ordering != Ordering::Less,
+                    ScalarFuncSig::EQInt => ordering == Ordering::Equal,
+                    ScalarFuncSig::NEInt => ordering != Ordering::Equal,
+                    _ => unreachable!(),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: Can it be `ScalarFuncSig::NullEQInt`?

File: src/coprocessor/dag/expr/compare.rs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+#[inline]
+fn cmp_i64_with_unsigned_flag(lhs: i64,
+                              lhs_unsigned: bool,
+                              rhs: i64,
+                              rhs_unsigned: bool)
+                              -> Ordering {
+    match (lhs_unsigned, rhs_unsigned) {
+        (true, true) | (false, false) => lhs.cmp(&rhs),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: This is wrong. For example, when comparing `u64::MAX` to `5u64`, this function will return `Ordering::Less`.

File: src/coprocessor/dag/expr/compare.rs [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+        (true, false) => {
+            if rhs < 0 || lhs as u64 > i64::MAX as u64 {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: You mean `lhs < 0`?

File: src/coprocessor/dag/expr/compare.rs [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                Ordering::Less
+            } else {
+                lhs.cmp(&rhs)
+            }
+        }
+    }
+}
+
+#[inline]
+fn cmp_f64(lhs: f64, rhs: f64) -> Ordering {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: Please consider reusing the function defined in `datum.rs`.

File: src/coprocessor/dag/expr/compare.rs [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+use tipb::expression::ScalarFuncSig;
+
+use coprocessor::codec::{mysql, Datum};
+use super::{StatementContext, Result, FnCall};
+
+impl FnCall {
+    pub fn compare_int(&self,
+                       ctx: &StatementContext,
+                       row: &[Datum],
+                       sig: ScalarFuncSig)
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/compare.rs]: I think just implement `compare_int`without `ScalarFuncSig` and return Ordering directly like `tidb` did would be more clear for reviewer?

File: src/coprocessor/dag/expr/compare.rs (hunk 3 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/compare.rs]: copyed from tidb, I think it's correct.

File: src/coprocessor/dag/expr/compare.rs [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+                       -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        match (lhs, rhs) {
+            (None, _) | (_, None) => {
+                if sig == ScalarFuncSig::NullEQInt {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: Doesn't `NullEqInt` mean `null == null`? And `2 == null` should be `false`.

File: src/coprocessor/dag/expr/compare.rs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                    ScalarFuncSig::LTInt => ordering == Ordering::Less,
+                    ScalarFuncSig::LEInt => ordering != Ordering::Greater,
+                    ScalarFuncSig::GTInt => ordering == Ordering::Greater,
+                    ScalarFuncSig::GEInt => ordering != Ordering::Less,
+                    ScalarFuncSig::NEInt => ordering != Ordering::Equal,
+                    ScalarFuncSig::EQInt | ScalarFuncSig::NullEQInt => ordering == Ordering::Equal,
+                    _ => unreachable!(),
+                } as i64))
+            }
+            _ => Ok(None),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: if `sig` is `NullEq`, and `lhs` is `None`, but `rhs` is `Some`, should return `Ok(false)`.

Comment (hicqu) [inline src/coprocessor/dag/expr/compare.rs]: Sorry, not carefully here.

**Pull request merged**